### ISSUE TITLE
False positive `invalid-argument-type` when passing a bounded class-scoped typevar to an `object` parameter of the same generic class

### ISSUE THREAD

**Issue by jenshnielsen** [astral-sh/ty#4303]
### Summary

In the following code I run into an unexpected type checking error

```python
class G[T: int]:
    def takes_object(self, x: object) -> bool:
        return True

    def caller(self, t: T) -> None:
        self.takes_object(t)
```

```
error[invalid-argument-type]: Argument to bound method `G.takes_object` is incorrect
 --> repro.py:6:27
  |
6 |         self.takes_object(t)
  |                           ^ Argument type `object` does not satisfy upper bound `int` of type variable `T`
info: Type variable defined here
 --> repro.py:1:9
  |
1 | class G[T: int]:
  |         ^^^^^^
```

This is surprizing since `self.takes_object` accepts any object so there should not be any issue accepting a T bound to int. I originally ran into this in a more complex class where `takes_object` was a `__contains__` method that accepts any object and returns False if called with a not T. 

I have looked through all issues matching a search of `object generic` and did not find any preexisting issues. 

### Version

0.0.72

**Comment by jenshnielsen** [comment]
Below is some AI exploration that may or may not be useful:

**Title:** False positive `invalid-argument-type` when passing a bounded class-scoped typevar to an `object` parameter of the same generic class

---

### Summary

Inside a generic class whose type parameter has an upper bound (or constraints), calling a method of
that same class with an argument whose type *is* that type parameter produces a false positive, but
only when the parameter is declared `object`:

```python
class G[T: int]:
    def takes_object(self, x: object) -> bool:
        return True

    def caller(self, t: T) -> None:
        self.takes_object(t)
```

```
error[invalid-argument-type]: Argument to bound method `G.takes_object` is incorrect
 --> repro.py:6:27
  |
6 |         self.takes_object(t)
  |                           ^ Argument type `object` does not satisfy upper bound `int` of type variable `T`
info: Type variable defined here
 --> repro.py:1:9
  |
1 | class G[T: int]:
  |         ^^^^^^
```

`T` is bounded by `int`, so any value of type `T` is an `int` and therefore an `object`. The call is
clearly valid, and both pyright and mypy accept it.

The reported argument type is `object`, not `T`, which suggests ty is re-solving the *class-scoped*
typevar `T` from the call site (taking the declared parameter type `object` as the solution) instead
of using the specialization already fixed by the receiver, and then validating that solution against
`T`'s own declared bound.

### Where this actually bites

The reason this is more than a curiosity: `Container.__contains__` is declared `(self, x: object)`,
so any bounded generic container that implements `__contains__` cannot use `in` on itself. This is
where I originally hit it:

```python
from collections.abc import Collection, Iterable, Iterator


class DictSet[P: int](Collection[P]):
    def __init__(self, items: Iterable[P] | None = None) -> None:
        self._dict: dict[P, None] = {}
        if items is not None:
            for item in items:
                self._dict[item] = None

    def __contains__(self, x: object) -> bool:
        return x in self._dict

    def __iter__(self) -> Iterator[P]:
        return iter(self._dict)

    def __len__(self) -> int:
        return len(self._dict)

    def issuperset(self, other: "DictSet[P]") -> bool:
        return all(item in self for item in other)
```

```
error[unsupported-operator]: Operator `in` is not supported between objects of type `P@DictSet` and `Self@issuperset`
 --> real.py:21:20
```

Spelling the call out as `self.__contains__(item)` gives the `invalid-argument-type` form of the
same error.

### Conditions

I narrowed down when it does and does not fire. It has nothing to do with `Collection`, iteration,
or comprehensions.

Triggers the false positive:

* The argument's type is the class's own type parameter `T`.
* The parameter is declared exactly `object`.
* `T` has an upper bound (`T: int`) **or** constraints (`T: (int, str)`); the constrained case
  reports ``Argument type `object` does not satisfy constraints (`int`, `str`) of type variable `T` ``.
* The receiver is `self`, an explicitly annotated `self: G[T]`, another parameter `other: G[T]`, or
  even a fully specialized `other: G[int]`.
* Instance methods and classmethods alike.

Does **not** trigger:

* Parameter declared `int`, `int | str`, `float`, `Any`, or `T`.
* An unbounded type parameter.
* Argument that is not typed as `T` (e.g. a literal `1`, or a separate `o: object` parameter).
* Receiver of a non-generic class, or of a *different* generic class (`Other[int]`).
* Staticmethods, or builtins such as `repr(t)` / `print(t)`.
* A function-scoped bounded typevar: `def f[T: int](g: G[int], t: T) -> None: g.takes_object(t)`.

Full matrix, self-contained and stdlib-only:

```python
from typing import Any


class Plain:
    def takes_object(self, x: object) -> bool: ...


class Other[S: int]:
    def takes_object(self, x: object) -> bool: ...


class G[T: int]:
    def takes_object(self, x: object) -> bool: ...
    def takes_int(self, x: int) -> bool: ...
    def takes_union(self, x: int | str) -> bool: ...
    def takes_any(self, x: Any) -> bool: ...
    def takes_T(self, x: T) -> bool: ...

    @staticmethod
    def static_takes_object(x: object) -> bool: ...

    @classmethod
    def cls_takes_object(cls, x: object) -> bool: ...

    # --- all five report invalid-argument-type ---
    def e1(self, t: T) -> None:
        self.takes_object(t)

    def e2(self: "G[T]", t: T) -> None:
        self.takes_object(t)

    def e3(self, other: "G[T]", t: T) -> None:
        other.takes_object(t)

    def e4(self, other: "G[int]", t: T) -> None:
        other.takes_object(t)

    def e5(self, t: T) -> None:
        self.cls_takes_object(t)

    # --- none of these report anything ---
    def ok1(self, t: T) -> None:
        self.takes_int(t)

    def ok2(self, t: T) -> None:
        self.takes_union(t)

    def ok3(self, t: T) -> None:
        self.takes_any(t)

    def ok4(self, t: T) -> None:
        self.takes_T(t)

    def ok5(self, t: T) -> None:
        self.takes_object(1)

    def ok6(self, o: object) -> None:
        self.takes_object(o)

    def ok7(self, t: T) -> None:
        G.static_takes_object(t)

    def ok8(self, p: Plain, t: T) -> None:
        p.takes_object(t)

    def ok9(self, o: Other[int], t: T) -> None:
        o.takes_object(t)

    def ok10(self, t: T) -> None:
        repr(t)
        print(t)


class Unbounded[T]:
    def takes_object(self, x: object) -> bool: ...

    def ok(self, t: T) -> None:  # fine
        self.takes_object(t)


class Constrained[T: (int, str)]:
    def takes_object(self, x: object) -> bool: ...

    def err(self, t: T) -> None:  # errors
        self.takes_object(t)


def free_typevar[T: int](g: G[int], t: T) -> None:  # fine
    g.takes_object(t)
```

(The `empty-body` errors in that snippet are expected and unrelated.)

That `object` is the only parameter type affected, while `int | str`, `float` (`int | float`) and
`Any` are all fine, makes me suspect this is related to top-materialization of a typevar not
respecting its bound — cf. #2660 / #666 — but I couldn't find an existing report of this symptom.

### Related but distinct

* #2467 — same diagnostic wording, but about `Self` and a `NewType` bound; closed as completed.
* #3871 — same diagnostic family, union bound, goes through a property setter.
* #1503 / #2585 — method calls on typevars constrained to a union of classes; different mechanism.
* #2652 — `unsupported-operator` for a `float`-bounded typevar; fixed, and the discussion there
  explicitly notes that an `int` bound works, which is the case that is broken here.

### Workarounds

For anyone hitting this, these all typecheck cleanly on 0.0.72:

```python
# route through the underlying container
return all(item in self._dict for item in other)

# or widen the element explicitly
return all(cast(object, item) in self for item in other)
```

Assigning `self` to a wider local (`me: Collection[int] = self`) does *not* help, since the inferred
type stays `Self@...`.

### Version

`ty 0.0.72`

**Comment by carljm** [comment]
Thanks! This is a bug, not specific to `object`. This also fails:

```py
class Base: pass
class Child(Base): pass

class G[T: Child]:
    def takes_base(self, x: Base) -> None: pass

    def caller(self, t: T) -> None:
        self.takes_base(t)
```

Root cause looks like a reversed polarity check in the constraint solver.

**Comment by carljm** [comment]
This is fixed by astral-sh/ruff#27653. I've opened [a PR](https://github.com/astral-sh/ruff/pull/27841) against that PR to add some specific tests for it.
